Src/FluidCaching/LifespanManager.cs

```diff
@@ -89,50 +98,61 @@ private void CleanUp(DateTime now)
                 int itemsAboveCapacity = Stats.Current - Stats.Capacity;
                 AgeBag<T> bag = bags[OldestBagIndex];
 
-                while (AlmostOutOfBags || bag.HasExpired(maxAge, now) ||
-                       (itemsAboveCapacity > 0 && bag.HasReachedMinimumAge(minAge, now)))
+                while (!HasProcessedAllBags && (AlmostOutOfBags || BagNeedsCleaning(bag, itemsAboveCapacity, now)))
                 {
-                    // cache is still too big / old so remove oldest bag
-                    Node<T> node = bag.First;
-                    bag.First = null;
+                    itemsAboveCapacity = CleanBag(bag, itemsAboveCapacity);
 
-                    while (node != null)
-                    {
-                        Node<T> next = node.Next;
-                        node.Next = null;
-                        if (node.Value != null && node.Bag != null)
-                        {
-                            if (node.Bag == bag)
-                            {
-                                // item has not been touched since bag was closed, so remove it from LifespanMgr
-                                ++itemsAboveCapacity;
-                                node.Remove();
-                            }
-                            else
-                            {
-                                // item has been touched and should be moved to correct age bag now
-                                node.Next = node.Bag.First;
-                                node.Bag.First = node;
-                            }
-                        }
+                    ++OldestBagIndex;
+                    bag = bags[OldestBagIndex];
+                }
 
-                        node = next;
-                    }
+                ++CurrentBagIndex;
+                OpenBag(CurrentBagIndex);
+
+                EnsureIndexIsValid();
+            }
+        }
+
+        private static int CleanBag(AgeBag<T> bag, int itemsAboveCapacity)
```

Comment:
Maybe it would be easier to understand, if the `itemsAboveCapacity` was passed by reference instead of returning the new value for it.